[quant][graphmode][fx] Add graph mode quantization on fx

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43335 [quant][graphmode][fx] Add support for batchnorm relu
* #43334 [quant][graphmode][fx] Add support for batchnorm
* #43333 [quant][graphmode][fx] Add support for cat
* #43332 [quant][graphmode][fx] Add support for add relu
* #43331 [quant][graphmode][fx] Add support for add
* #43287 [quant][graphmode][fx] Add support for conv module + relu
* #43286 [quant] Make OP_LIST_TO_FUSER_METHOD public
* #43285 [quant][graphmode][fx] Add support for conv module
* #43229 [quant][graphmode][fx][test] Add per op test for graph mode quant on fx
* **#43175 [quant][graphmode][fx] Add graph mode quantization on fx**

Summary:
This PR added graph mode quantization on fx: https://github.com/pytorch/pytorch/pull/42741
Currently it matches eager mode quantization for torchvision with static/dynamic/qat
ddp/synbn test is still wip

File structure
```
torch/quantization
        - _quantize_fx.py - top level APIs, currently Quantizer and fuse
        - fx - implementations
          - fuse.py - fuse conv bn, similar to eager mode fusion but also works for functional relu
          - quantize.py - implementation for Quantizer, static/qat/dynamic patterns are registered here as well
          - pattern_utils.py - utility function for pattern registration
          - utils.py - other utility functions

test/quantization/test_quantize_fx.py -- test for fx based graph mode quantization
```
Note that the current API is not final.

Next Step:
- Add support for all quantized ops that's implemented right now: https://github.com/pytorch/pytorch/tree/master/aten/src/ATen/native/quantized/cpu

Test Plan:
python test/test_quantization.py TestQuantizeFx

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23178602](https://our.internmc.facebook.com/intern/diff/D23178602)

jerryzh168 committed:
[quant][graphmode][fx] Add graph mode quantization on fx

Summary:
This PR added graph mode quantization on fx: https://github.com/pytorch/pytorch/pull/42741
Currently it matches eager mode quantization for torchvision with static/dynamic/qat
ddp/synbn test is still wip

Test Plan:
python test/test_quantization.py TestQuantizeFx

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add graph mode quantization on fx"


Summary:
This PR added graph mode quantization on fx: https://github.com/pytorch/pytorch/pull/42741
Currently it matches eager mode quantization for torchvision with static/dynamic/qat
ddp/synbn test is still wip

Test Plan:
python test/test_quantization.py TestQuantizeFx

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23178602](https://our.internmc.facebook.com/intern/diff/D23178602)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add graph mode quantization on fx"


Summary:
This PR added graph mode quantization on fx: https://github.com/pytorch/pytorch/pull/42741
Currently it matches eager mode quantization for torchvision with static/dynamic/qat
ddp/synbn test is still wip

Test Plan:
python test/test_quantization.py TestQuantizeFx

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23178602](https://our.internmc.facebook.com/intern/diff/D23178602)

[ghstack-poisoned]

vkuzo reviewed (approved):
awesomesauce

we can iterate in future PRs.  One thing which would be great to add is docblocks on public APIs, and a design document for the quantization integration on top of GraphModule

vkuzo commented on test/quantization/test_quantize_fx.py:
```diff
@@ -0,0 +1,116 @@
+)
+
+class TestQuantizeFx(QuantizationTestCase):
+    def test_functional(self):
```
would be awesome to also have a basic QAT test case

jerryzh168 replied:
currently I only have integration tests with torchvision, plan to add it after ddp test works. 
also it will be added in the future when we add individual op tests as well.

vkuzo commented on torch/quantization/fx/fuse.py:
```diff
@@ -0,0 +1,155 @@
+import copy
+
+# Fusion Patterns
+@register_fusion_pattern((torch.nn.BatchNorm2d, torch.nn.Conv2d))
```
haven't looked at how this is used, but one question I have is if we can make the order of the pattern follow the order of the layers in the model (i.e. Conv-BN instead of BN-Conv)

jerryzh168 replied:
that is doable, but the matching is from later ops to former ops, it will confusing when people add new fusion patterns, I feel it would be better to keep the current order considering how matching works right now.

vkuzo replied:
cool, then maybe just a comment explaining this would work

jerryzh168 replied:
had a brief description in the quantize file, I'll move it to pattern_utils.py

vkuzo replied:
wanted to follow up on this, we should discuss defining patterns in the order that they occur in the graph, and changing the matching logic.  Reading through the current implementation, it's not intuitive that the order is reversed.

jerryzh168 replied:
sure, I'll change the order

vkuzo commented on torch/quantization/fx/quantize.py:
```diff
@@ -0,0 +1,765 @@
+@register_quant_pattern(operator.add)
+@register_quant_pattern((torch.nn.ReLU, operator.add))
+@register_quant_pattern((torch.nn.functional.relu, operator.add))
+class Add(QuantizerPattern):
```
I think for all of these an example (input X, output Y) would be helpful for people to understand the API without having to read every single line of code

jerryzh168 replied:
Sure

jerryzh168 replied:
will probably do this in the end after everything is done, since the code might be changing.

vkuzo replied:
I think it's important to do this sooner than later, because the code is harder to read than it needs to be without the docs.  Adding documentation now and adjusting it when things change should be a net benefit, because engineers will spend time reading this code.

jerryzh168 replied:
Sure

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add graph mode quantization on fx"


Summary:
This PR added graph mode quantization on fx: https://github.com/pytorch/pytorch/pull/42741
Currently it matches eager mode quantization for torchvision with static/dynamic/qat
ddp/synbn test is still wip

File structure
```
torch/quantization
        - _quantize_fx.py - top level APIs
        - fx - implementations
          - fuse.py - fuse conv bn
          - quantize.py - implementation for Quantizer
          - pattern_utils.py - utility function for patterns
          - utils.py - other utility functions

test/quantization/test_quantize_fx.py -- test for fx based graph mode quantization
```
Notice that the current API is not final.

Test Plan:
python test/test_quantization.py TestQuantizeFx

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23178602](https://our.internmc.facebook.com/intern/diff/D23178602)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add graph mode quantization on fx"


Summary:
This PR added graph mode quantization on fx: https://github.com/pytorch/pytorch/pull/42741
Currently it matches eager mode quantization for torchvision with static/dynamic/qat
ddp/synbn test is still wip

File structure
```
torch/quantization
        - _quantize_fx.py - top level APIs, currently Quantizer and fuse
        - fx - implementations
          - fuse.py - fuse conv bn, similar to eager mode fusion but also works for functional relu
          - quantize.py - implementation for Quantizer, static/qat/dynamic patterns are registered here as well
          - pattern_utils.py - utility function for pattern registration
          - utils.py - other utility functions

test/quantization/test_quantize_fx.py -- test for fx based graph mode quantization
```
Notice that the current API is not final.

Next Step:
- Add support for all quantized ops that's implemented right now: https://github.com/pytorch/pytorch/tree/master/aten/src/ATen/native/quantized/cpu

Test Plan:
python test/test_quantization.py TestQuantizeFx

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23178602](https://our.internmc.facebook.com/intern/diff/D23178602)

[ghstack-poisoned]

supriyar commented on torch/quantization/fx/fuse.py:
```diff
@@ -0,0 +1,155 @@
+
+        return match_map
+
+def fuse(graph_module, inplace=False):
```
Where do we call the `fuse()` function?

jerryzh168 replied:
before quantization, there is an example in test/quantization/test_quantize_fx.py

supriyar commented on torch/quantization/fx/quantize.py:
```diff
@@ -0,0 +1,765 @@
+# 2. Post Training Dynamic Quantizatoin Patterns
+@register_dynamic_pattern(torch.nn.Linear)
+@register_dynamic_pattern(torch.nn.functional.linear)
+class DynamicLinear(QuantizerPattern):
```
For dynamic quant, can we run subset of the graph to run only weight observers?

jerryzh168 replied:
This is not handled yet, but should still be doable.

jerryzh168 commented on torch/quantization/fx/quantize.py:
```diff
@@ -0,0 +1,727 @@
+                qdynamic_linear_args = [linear_input, packed_weight]
+                return quantizer.quantized_graph.call_function(torch.ops.quantized.linear_dynamic, qdynamic_linear_args, kwargs)
+
+class Quantizer:
```
since these pattern also contain implementations of the feature I feel it might make more sense to put them together. see reference to `quantizer` in the `convert` function of the pattern. the pattern can be considered to be 'tied' to the quantizer implementation.

supriyar replied:
Should we consider calling it something other than `Quantizer`? Sine the quantization backend currently uses it for creating different types of qtensors.

jerryzh168 replied:
yeah, I think so. the official API is not figured out yet, I think maybe we'll not expose this to user and have a quantize_fx as a top level function in the end.

raghuramank100 replied:
Second comments here: Request moving the quantizer to a separate file and calling it something else other than Quantizer.

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add graph mode quantization on fx"


Summary:
This PR added graph mode quantization on fx: https://github.com/pytorch/pytorch/pull/42741
Currently it matches eager mode quantization for torchvision with static/dynamic/qat
ddp/synbn test is still wip

File structure
```
torch/quantization
        - _quantize_fx.py - top level APIs, currently Quantizer and fuse
        - fx - implementations
          - fuse.py - fuse conv bn, similar to eager mode fusion but also works for functional relu
          - quantize.py - implementation for Quantizer, static/qat/dynamic patterns are registered here as well
          - pattern_utils.py - utility function for pattern registration
          - utils.py - other utility functions

test/quantization/test_quantize_fx.py -- test for fx based graph mode quantization
```
Notice that the current API is not final.

Next Step:
- Add support for all quantized ops that's implemented right now: https://github.com/pytorch/pytorch/tree/master/aten/src/ATen/native/quantized/cpu

Test Plan:
python test/test_quantization.py TestQuantizeFx

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23178602](https://our.internmc.facebook.com/intern/diff/D23178602)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add graph mode quantization on fx"


Summary:
This PR added graph mode quantization on fx: https://github.com/pytorch/pytorch/pull/42741
Currently it matches eager mode quantization for torchvision with static/dynamic/qat
ddp/synbn test is still wip

File structure
```
torch/quantization
        - _quantize_fx.py - top level APIs, currently Quantizer and fuse
        - fx - implementations
          - fuse.py - fuse conv bn, similar to eager mode fusion but also works for functional relu
          - quantize.py - implementation for Quantizer, static/qat/dynamic patterns are registered here as well
          - pattern_utils.py - utility function for pattern registration
          - utils.py - other utility functions

test/quantization/test_quantize_fx.py -- test for fx based graph mode quantization
```
Notice that the current API is not final.

Next Step:
- Add support for all quantized ops that's implemented right now: https://github.com/pytorch/pytorch/tree/master/aten/src/ATen/native/quantized/cpu

Test Plan:
python test/test_quantization.py TestQuantizeFx

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23178602](https://our.internmc.facebook.com/intern/diff/D23178602)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add graph mode quantization on fx"


Summary:
This PR added graph mode quantization on fx: https://github.com/pytorch/pytorch/pull/42741
Currently it matches eager mode quantization for torchvision with static/dynamic/qat
ddp/synbn test is still wip

File structure
```
torch/quantization
        - _quantize_fx.py - top level APIs, currently Quantizer and fuse
        - fx - implementations
          - fuse.py - fuse conv bn, similar to eager mode fusion but also works for functional relu
          - quantize.py - implementation for Quantizer, static/qat/dynamic patterns are registered here as well
          - pattern_utils.py - utility function for pattern registration
          - utils.py - other utility functions

test/quantization/test_quantize_fx.py -- test for fx based graph mode quantization
```
Notice that the current API is not final.

Next Step:
- Add support for all quantized ops that's implemented right now: https://github.com/pytorch/pytorch/tree/master/aten/src/ATen/native/quantized/cpu

Test Plan:
python test/test_quantization.py TestQuantizeFx

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23178602](https://our.internmc.facebook.com/intern/diff/D23178602)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add graph mode quantization on fx"


Summary:
This PR added graph mode quantization on fx: https://github.com/pytorch/pytorch/pull/42741
Currently it matches eager mode quantization for torchvision with static/dynamic/qat
ddp/synbn test is still wip

File structure
```
torch/quantization
        - _quantize_fx.py - top level APIs, currently Quantizer and fuse
        - fx - implementations
          - fuse.py - fuse conv bn, similar to eager mode fusion but also works for functional relu
          - quantize.py - implementation for Quantizer, static/qat/dynamic patterns are registered here as well
          - pattern_utils.py - utility function for pattern registration
          - utils.py - other utility functions

test/quantization/test_quantize_fx.py -- test for fx based graph mode quantization
```
Notice that the current API is not final.

Next Step:
- Add support for all quantized ops that's implemented right now: https://github.com/pytorch/pytorch/tree/master/aten/src/ATen/native/quantized/cpu

Test Plan:
python test/test_quantization.py TestQuantizeFx

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23178602](https://our.internmc.facebook.com/intern/diff/D23178602)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add graph mode quantization on fx"


Summary:
This PR added graph mode quantization on fx: https://github.com/pytorch/pytorch/pull/42741
Currently it matches eager mode quantization for torchvision with static/dynamic/qat
ddp/synbn test is still wip

File structure
```
torch/quantization
        - _quantize_fx.py - top level APIs, currently Quantizer and fuse
        - fx - implementations
          - fuse.py - fuse conv bn, similar to eager mode fusion but also works for functional relu
          - quantize.py - implementation for Quantizer, static/qat/dynamic patterns are registered here as well
          - pattern_utils.py - utility function for pattern registration
          - utils.py - other utility functions

test/quantization/test_quantize_fx.py -- test for fx based graph mode quantization
```
Notice that the current API is not final.

Next Step:
- Add support for all quantized ops that's implemented right now: https://github.com/pytorch/pytorch/tree/master/aten/src/ATen/native/quantized/cpu

Test Plan:
python test/test_quantization.py TestQuantizeFx

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23178602](https://our.internmc.facebook.com/intern/diff/D23178602)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add graph mode quantization on fx"


Summary:
This PR added graph mode quantization on fx: https://github.com/pytorch/pytorch/pull/42741
Currently it matches eager mode quantization for torchvision with static/dynamic/qat
ddp/synbn test is still wip

File structure
```
torch/quantization
        - _quantize_fx.py - top level APIs, currently Quantizer and fuse
        - fx - implementations
          - fuse.py - fuse conv bn, similar to eager mode fusion but also works for functional relu
          - quantize.py - implementation for Quantizer, static/qat/dynamic patterns are registered here as well
          - pattern_utils.py - utility function for pattern registration
          - utils.py - other utility functions

test/quantization/test_quantize_fx.py -- test for fx based graph mode quantization
```
Notice that the current API is not final.

Next Step:
- Add support for all quantized ops that's implemented right now: https://github.com/pytorch/pytorch/tree/master/aten/src/ATen/native/quantized/cpu

Test Plan:
python test/test_quantization.py TestQuantizeFx

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23178602](https://our.internmc.facebook.com/intern/diff/D23178602)

[ghstack-poisoned]

jerryzh168 commented:
cc @jamesr66a @zdevito FYI

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add graph mode quantization on fx"


Summary:
This PR added graph mode quantization on fx: https://github.com/pytorch/pytorch/pull/42741
Currently it matches eager mode quantization for torchvision with static/dynamic/qat
ddp/synbn test is still wip

File structure
```
torch/quantization
        - _quantize_fx.py - top level APIs, currently Quantizer and fuse
        - fx - implementations
          - fuse.py - fuse conv bn, similar to eager mode fusion but also works for functional relu
          - quantize.py - implementation for Quantizer, static/qat/dynamic patterns are registered here as well
          - pattern_utils.py - utility function for pattern registration
          - utils.py - other utility functions

test/quantization/test_quantize_fx.py -- test for fx based graph mode quantization
```
Note that the current API is not final.

Next Step:
- Add support for all quantized ops that's implemented right now: https://github.com/pytorch/pytorch/tree/master/aten/src/ATen/native/quantized/cpu

Test Plan:
python test/test_quantization.py TestQuantizeFx

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23178602](https://our.internmc.facebook.com/intern/diff/D23178602)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add graph mode quantization on fx"


Summary:
This PR added graph mode quantization on fx: https://github.com/pytorch/pytorch/pull/42741
Currently it matches eager mode quantization for torchvision with static/dynamic/qat
ddp/synbn test is still wip

File structure
```
torch/quantization
        - _quantize_fx.py - top level APIs, currently Quantizer and fuse
        - fx - implementations
          - fuse.py - fuse conv bn, similar to eager mode fusion but also works for functional relu
          - quantize.py - implementation for Quantizer, static/qat/dynamic patterns are registered here as well
          - pattern_utils.py - utility function for pattern registration
          - utils.py - other utility functions

test/quantization/test_quantize_fx.py -- test for fx based graph mode quantization
```
Note that the current API is not final.

Next Step:
- Add support for all quantized ops that's implemented right now: https://github.com/pytorch/pytorch/tree/master/aten/src/ATen/native/quantized/cpu

Test Plan:
python test/test_quantization.py TestQuantizeFx

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23178602](https://our.internmc.facebook.com/intern/diff/D23178602)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add graph mode quantization on fx"


Summary:
This PR added graph mode quantization on fx: https://github.com/pytorch/pytorch/pull/42741
Currently it matches eager mode quantization for torchvision with static/dynamic/qat
ddp/synbn test is still wip

File structure
```
torch/quantization
        - _quantize_fx.py - top level APIs, currently Quantizer and fuse
        - fx - implementations
          - fuse.py - fuse conv bn, similar to eager mode fusion but also works for functional relu
          - quantize.py - implementation for Quantizer, static/qat/dynamic patterns are registered here as well
          - pattern_utils.py - utility function for pattern registration
          - utils.py - other utility functions

test/quantization/test_quantize_fx.py -- test for fx based graph mode quantization
```
Note that the current API is not final.

Next Step:
- Add support for all quantized ops that's implemented right now: https://github.com/pytorch/pytorch/tree/master/aten/src/ATen/native/quantized/cpu

Test Plan:
python test/test_quantization.py TestQuantizeFx

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23178602](https://our.internmc.facebook.com/intern/diff/D23178602)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add graph mode quantization on fx"


Summary:
This PR added graph mode quantization on fx: https://github.com/pytorch/pytorch/pull/42741
Currently it matches eager mode quantization for torchvision with static/dynamic/qat
ddp/synbn test is still wip

File structure
```
torch/quantization
        - _quantize_fx.py - top level APIs, currently Quantizer and fuse
        - fx - implementations
          - fuse.py - fuse conv bn, similar to eager mode fusion but also works for functional relu
          - quantize.py - implementation for Quantizer, static/qat/dynamic patterns are registered here as well
          - pattern_utils.py - utility function for pattern registration
          - utils.py - other utility functions

test/quantization/test_quantize_fx.py -- test for fx based graph mode quantization
```
Note that the current API is not final.

Next Step:
- Add support for all quantized ops that's implemented right now: https://github.com/pytorch/pytorch/tree/master/aten/src/ATen/native/quantized/cpu

Test Plan:
python test/test_quantization.py TestQuantizeFx

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23178602](https://our.internmc.facebook.com/intern/diff/D23178602)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add graph mode quantization on fx"


Summary:
This PR added graph mode quantization on fx: https://github.com/pytorch/pytorch/pull/42741
Currently it matches eager mode quantization for torchvision with static/dynamic/qat
ddp/synbn test is still wip

File structure
```
torch/quantization
        - _quantize_fx.py - top level APIs, currently Quantizer and fuse
        - fx - implementations
          - fuse.py - fuse conv bn, similar to eager mode fusion but also works for functional relu
          - quantize.py - implementation for Quantizer, static/qat/dynamic patterns are registered here as well
          - pattern_utils.py - utility function for pattern registration
          - utils.py - other utility functions

test/quantization/test_quantize_fx.py -- test for fx based graph mode quantization
```
Note that the current API is not final.

Next Step:
- Add support for all quantized ops that's implemented right now: https://github.com/pytorch/pytorch/tree/master/aten/src/ATen/native/quantized/cpu

Test Plan:
python test/test_quantization.py TestQuantizeFx

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23178602](https://our.internmc.facebook.com/intern/diff/D23178602)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add graph mode quantization on fx"


Summary:
This PR added graph mode quantization on fx: https://github.com/pytorch/pytorch/pull/42741
Currently it matches eager mode quantization for torchvision with static/dynamic/qat
ddp/synbn test is still wip

File structure
```
torch/quantization
        - _quantize_fx.py - top level APIs, currently Quantizer and fuse
        - fx - implementations
          - fuse.py - fuse conv bn, similar to eager mode fusion but also works for functional relu
          - quantize.py - implementation for Quantizer, static/qat/dynamic patterns are registered here as well
          - pattern_utils.py - utility function for pattern registration
          - utils.py - other utility functions

test/quantization/test_quantize_fx.py -- test for fx based graph mode quantization
```
Note that the current API is not final.

Next Step:
- Add support for all quantized ops that's implemented right now: https://github.com/pytorch/pytorch/tree/master/aten/src/ATen/native/quantized/cpu

Test Plan:
python test/test_quantization.py TestQuantizeFx

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23178602](https://our.internmc.facebook.com/intern/diff/D23178602)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add graph mode quantization on fx"


Summary:
This PR added graph mode quantization on fx: https://github.com/pytorch/pytorch/pull/42741
Currently it matches eager mode quantization for torchvision with static/dynamic/qat
ddp/synbn test is still wip

File structure
```
torch/quantization
        - _quantize_fx.py - top level APIs, currently Quantizer and fuse
        - fx - implementations
          - fuse.py - fuse conv bn, similar to eager mode fusion but also works for functional relu
          - quantize.py - implementation for Quantizer, static/qat/dynamic patterns are registered here as well
          - pattern_utils.py - utility function for pattern registration
          - utils.py - other utility functions

test/quantization/test_quantize_fx.py -- test for fx based graph mode quantization
```
Note that the current API is not final.

Next Step:
- Add support for all quantized ops that's implemented right now: https://github.com/pytorch/pytorch/tree/master/aten/src/ATen/native/quantized/cpu

Test Plan:
python test/test_quantization.py TestQuantizeFx

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23178602](https://our.internmc.facebook.com/intern/diff/D23178602)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Add graph mode quantization on fx"


Summary:
This PR added graph mode quantization on fx: https://github.com/pytorch/pytorch/pull/42741
Currently it matches eager mode quantization for torchvision with static/dynamic/qat
ddp/synbn test is still wip

File structure
```
torch/quantization
        - _quantize_fx.py - top level APIs, currently Quantizer and fuse
        - fx - implementations
          - fuse.py - fuse conv bn, similar to eager mode fusion but also works for functional relu
          - quantize.py - implementation for Quantizer, static/qat/dynamic patterns are registered here as well
          - pattern_utils.py - utility function for pattern registration
          - utils.py - other utility functions

test/quantization/test_quantize_fx.py -- test for fx based graph mode quantization
```
Note that the current API is not final.

Next Step:
- Add support for all quantized ops that's implemented right now: https://github.com/pytorch/pytorch/tree/master/aten/src/ATen/native/quantized/cpu

Test Plan:
python test/test_quantization.py TestQuantizeFx

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23178602](https://our.internmc.facebook.com/intern/diff/D23178602)

[ghstack-poisoned]
